acm policies for dedicated-admin and cluster admin rbac

lnguyen1401 opened the pull request:
### What type of PR is this?
Adding SelectorSyncSet 

### What this PR does / why we need it?
Create SelectorSyncSet that would land the ACM policies for dedicated-admin and cluster-admin rbac on the Hypershift service cluster
### Which Jira/Github issue(s) this PR fixes?
https://issues.redhat.com/browse/OSD-13407
_Fixes #_

### Special notes for your reviewer:
Suggested PR test: 
- Pull the PR
- Remove these 2 files 
deploy/acm-policies/generated-policies/00.cluster-admin-rolebinding.policy.yaml
deploy/acm-policies/generated-policies/00.dedicated-admin-config.policy.yaml
- run make
this should regenerate the 2 policies files above and add them to SelectorSyncSet

Policy Generator: https://github.com/stolostron/policy-generator-plugin
Doc on Generating Policy: https://cloud.redhat.com/blog/generating-governance-policies-using-kustomize-and-gitops
More Doc on working with policies: https://docs.google.com/document/d/1ZsaaY3KcdUyZp-Nn5fVVmsYRFG8y4cdHPO4FoCFUmGw/edit# 

Deployements included in the each policy: 

policy-cluster-admin: https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/osd-cluster-admin/03-cluster-admin.ClusterRoleBinding.yaml

 policy-dedicated-admin:
+ [03-dedicated-admins-cluster.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/03-dedicated-admins-cluster.ClusterRole.yaml)
+ [03-dedicated-admins-project.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/03-dedicated-admins-project.ClusterRole.yaml)
+ [05-dedicated-admins-aggregate-cluster.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/05-dedicated-admins-aggregate-cluster.ClusterRole.yaml)
+ [05-dedicated-admins-aggregate-project.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/05-dedicated-admins-aggregate-project.ClusterRole.yaml)
+ [05-dedicated-readers.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/05-dedicated-readers.ClusterRole.yaml)
+ [05-osd-readers.ClusterRole.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/05-osd-readers.ClusterRole.yaml)
+ [07-dedicated-admins-cluster.ClusterRoleBinding.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/07-dedicated-admins-cluster.ClusterRoleBinding.yaml)
+ [10-dedicated-admins-operators.Role.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/10-dedicated-admins-operators.Role.yaml)
+ [10-dedicated-admins-operators.RoleBinding.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/10-dedicated-admins-operators.RoleBinding.yaml)
+ [20-dedicated-admins-marketplace.Role.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/20-dedicated-admins-marketplace.Role.yaml)
+ [20-dedicated-admins-marketplace.RoleBinding.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/20-dedicated-admins-marketplace.RoleBinding.yaml)
+ [30-dedicated-admins-dns.Role.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/30-dedicated-admins-dns.Role.yaml)
+ [30-dedicated-admins-dns.RoleBinding.yaml](https://github.com/openshift/managed-cluster-config/blob/d251fa4bb9652c35d33898b5e2c8adc8efc851f1/deploy/rbac-permissions-operator-config/30-dedicated-admins-dns.RoleBinding.yaml)

### Pre-checks (if applicable):
- [x] Tested latest changes against a cluster
- [x] Included documentation changes with PR
- [x] If this is a new object that is not intended for the FedRAMP environment (if unsure, please reach out to team FedRAMP), please exclude it with:

    ```yaml
    matchExpressions:
    - key: api.openshift.com/fedramp
      operator: NotIn
      values: ["true"]
    ```

jewzaam commented:
/hold

comments to come in a bit.

jewzaam commented:
How is the policy created?  Missing policy generation tooling, docs, etc.

How do I know when a PR changes either the Policy or the original roles that everything is consistent?  Missing PR check.

Recommend use only `matchExpressions` and not mix both `matchExpressions` and `matchLabels` in the same config.

Namespace name is vague, "acm-policies" tell me it's policy but not much more.  Recommend a more meaningful name.  Recommend using "openshift-" prefix for namespace so it is treated as a platform namespace and access is limited.

lnguyen1401 commented:
Updated PR including:

- Add Generating tool and Doc
- Changed the namespace for Policy from 'acm-policies' to 'openshift-rbac-policies'
- Only use matchExpression

lnguyen1401 commented:
Testing done on hypershift's service cluster in staging:
- Log into the service cluster in stage
- Apply the policies 
- result: 
` 
oc get policy -A
NAMESPACE                          NAME                                    REMEDIATION ACTION   COMPLIANCE STATE   AGE
1vikdtls3pig06ja3bglm7rmqe45p9gd   openshift-rbac-policies.cluster-admin                                           28s
1vj620mb94v1pt99beru1dd0n6bu3kfi   openshift-rbac-policies.cluster-admin                                           28s
1vjak0vj8d79oiu38ga2du8a71873g40   openshift-rbac-policies.cluster-admin                        Compliant          28s
1vp9qvh1m6b6hn21a3791cm3g7lev4o9   openshift-rbac-policies.cluster-admin                                           28s
1vqlma2953cajqg8ab6u1uh7c1v3feep   openshift-rbac-policies.cluster-admin                        Compliant          28s
1vr934lph13umpaqka03fvdhqaes32sp   openshift-rbac-policies.cluster-admin                                           28s
openshift-rbac-policies            cluster-admin                                                                   29s
`
- Cleaned up the testing in service cluster.

lnguyen1401 commented:
Currently i'm not sure how to do PR check for this MCC pr. @jewzaam can you give me some pointers? Thank you !

jewzaam commented:
@lnguyen1401 you probably don't need to change the pr check.  Right now script `hack/app_sre_pr_check.sh` will run and check that nothing changed when `make` is executed.  Policy generation should happen as a part of the default make target.  Therefore if this is not run the PR check will catch this already by the files changing.

iamkirkbater commented:
@lnguyen1401 @jewzaam  - is there anything that's preventing this from being deployed at this point? It looks like all of Naveen's concerns were addressed, but I want to double check.

Thanks!

lnguyen1401 commented:
@iamkirkbater No blockers for the PR from my end

fahlmant commented on deploy/acm-policies/policy-dedicated-admin.yaml:
```diff
@@ -0,0 +1,1019 @@
+                                      values:
+                                        - openshift-cluster-monitoring
+                                        - olm-operators
+                                        - cloud-ingress-operator
```
cloud-ingress-operator isn't present on hypershift clusters, so this could likely be removed. Same for some of the other values on this list and the namespace list.

lnguyen1401 replied:
I have removed cloud-ingress-operator,openshift-velero, rbac-permission-operator from this list and namespace list.

lnguyen1401 commented on deploy/acm-policies/policy-cluster-admin.yaml:
```diff
@@ -0,0 +1,63 @@
+kind: Policy
+metadata:
+    annotations:
+        policy.open-cluster-management.io/categories: CM Configuration Management
+        policy.open-cluster-management.io/controls: CM-2 Baseline Configuration
+        policy.open-cluster-management.io/standards: NIST SP 800-53
```
Yes, those are the default values used by the Generator : https://github.com/stolostron/policy-generator-plugin/blob/669d8aecde8f2ae44db6b9d21298f6e2396f5121/docs/policygenerator-reference.yaml#L24

boranx commented on deploy/acm-policies/policy-cluster-admin.yaml:
```diff
@@ -0,0 +1,63 @@
+                name: cluster-admin
+            spec:
+                object-templates:
+                    - complianceType: musthave
```
just for knowledge: is this creating the cluster-admin(itself and its dependent resources) if not exist? 
Or it's just for throwing a compliance alert that the policy didn't meet the expected criteria?

lnguyen1401 replied:
Yeah the Policy will Create the clusterrolebinding if it didn't exist already. if the complianceType set to inform, it would tell us that the clusterrolebinding is missing without creating it.

iamkirkbater commented:
@fahlmant - This LGTM at this point.  Do you have any further concerns before getting this merged?

rafael-azevedo commented:
cc @jewzaam is there anything still blocking the merging of this one ?

jewzaam reviewed (changes requested):
> How is the policy created? Missing policy generation tooling, docs, etc.

This is still not addressed.  There is no policy generation included.  When we change RBAC in other place how is this new acm-policies being kept up to date!?  This should be incorporated into the default make goal for the repository.

> How do I know when a PR changes either the Policy or the original roles that everything is consistent? Missing PR check.

Given the generation needs to be handled in the default make target, we get the PR check by default.  This needs validated.

---------

The files do not follow standard naming for MCC: ##-descriptiveName.Kind.yaml

lnguyen1401 commented:
Updated the PR to:
- Added policy generation to make default target
- Updated README for running instructions
- Changed policy naming convention to ##.<$detailed-name>.policy.yaml

lnguyen1401 commented:
/retest-required

lnguyen1401 commented:
Hello @jewzaam , How should i carry out this requirement ? 
`Given the generation needs to be handled in the default make target, we get the PR check by default. This needs validated.
`

iamkirkbater commented on Makefile:
```diff
@@ -57,3 +60,17 @@ generate-hive-templates: generate-oauth-templates
 		${GEN_TEMPLATE}; \
 	fi
+
+.PHONY: generate-acm-policies
+generate-acm-policies: generate-acm-policies
```
Is this a duplicate or am I just having a hard time reading this?

lnguyen1401 replied:
My first time working with makefile so i wasn't sure. I'll fix that !

jewzaam reviewed (changes requested):
The challenge is the acm-policies directory is already behind!  Changes have happened in the dedicated-admins-aggregate-cluster ClusterRole and dedicated-admins-marketplace Role an the addition of 4 SubjectPermission resources.  By creating a resource specific to acm policy generation we now have to maintain the same resource in two places.  Instead you can reference the existing resource.  See suggestions in policy-generator-config.yaml

---

I don't think we have a case of needing to exclude a resource but if there was a need we could split it out to a separate directory that is not referenced by policy generation.

---

There are some things that have been removed in the acm-policies but those can be left from what I'm seeing.  Meaning the additions won't hurt anything, they are conditions in rule aggregation that won't be triggered.  If they're cleaned up some day it will be because we know the fleet overall doesn't need it anywhere.

---

The `policy-generator-config.yaml` need to move out of `deploy/` as they're being bundled into the hive templates.  They can be put somewhere else like `resources/`.

lnguyen1401 commented on deploy/acm-policies/cluster-admins/policy-generator-config.yaml:
```diff
@@ -0,0 +1,16 @@
+policies:
+    - name: cluster-admin
+      manifests:
+          - path: ./deploy/acm-policies/cluster-admins/cluster-admin-role-binding.yaml
```
i changed it to ./deploy/osd-cluster-admin/03-cluster-admin.ClusterRoleBinding.yaml because the directory also contains config.yaml . would that be ok ?

jewzaam commented on Makefile:
```diff
@@ -53,7 +56,9 @@ generate-rosa-brand-logo:
 generate-hive-templates: generate-oauth-templates
 	if [ -z ${IN_CONTAINER} ]; then \
 		$(CONTAINER_ENGINE) pull quay.io/app-sre/python:3 && $(CONTAINER_ENGINE) tag quay.io/app-sre/python:3 python:3 || true; \
+		$(CONTAINER_ENGINE) run $(CONTAINER_RUN_FLAGS) python:3 /bin/sh -c "wget -nc https://github.com/stolostron/policy-generator-plugin/releases/download/v1.9.1/linux-amd64-PolicyGenerator; chmod +x linux-amd64-PolicyGenerator ; PolicyGenerator=$(shell pwd)/linux-amd64-PolicyGenerator ${GEN_POLICY}";\
```
Updates to not download to local filesystem by using `/tmp` and set `-u` to get correct file premissions in generated output.

```suggestion
		$(CONTAINER_ENGINE) run $(CONTAINER_RUN_FLAGS) -u `id -u ${USER}`:`id -g ${USER}` python:3 /bin/sh -c "wget -nc https://github.com/stolostron/policy-generator-plugin/releases/download/v1.9.1/linux-amd64-PolicyGenerator -O /tmp/linux-amd64-PolicyGenerator; chmod +x /tmp/linux-amd64-PolicyGenerator ; PolicyGenerator=/tmp/linux-amd64-PolicyGenerator ${GEN_POLICY}";\
```

lnguyen1401 replied:
Used this `--user $(id -u):$(id -g)` and it seems to work now. i'm not 100% sure it's the same as `-u `id -u ${USER}`:`id -g ${USER}`` in the suggetion

jewzaam reviewed (changes requested):
Not seeing a reason for separate directory for the namespace.  Recommend:
```
rm deploy/openshift-managed-rbac-policies/config.yaml
mv deploy/openshift-managed-rbac-policies/policy-namespace.yaml deploy/acm-policies/00-openshift-managed-rbac-policies.Namepsace.yaml
```

jewzaam reviewed (changes requested):
One suggestion for file permission fix.

Bigger problem is SubjectPermissions.  Those need to be converted to a policy.  This means enumerating all files in the PolicyGenerator so they are not pulled in and to somehow get generate-policy.sh to handle these SubjectPermissions.  They were added recently to the `deploy/rbac-permissions-operator-config` directory.

jewzaam commented on Makefile:
```diff
@@ -53,7 +59,10 @@ generate-rosa-brand-logo:
 generate-hive-templates: generate-oauth-templates
 	if [ -z ${IN_CONTAINER} ]; then \
 		$(CONTAINER_ENGINE) pull quay.io/app-sre/python:3 && $(CONTAINER_ENGINE) tag quay.io/app-sre/python:3 python:3 || true; \
+		$(CONTAINER_ENGINE) run $(CONTAINER_RUN_FLAGS) --user $(id -u):$(id -g) python:3 /bin/sh -c "cd `pwd -P`; pip install oyaml; ${GEN_POLICY_CONFIG}; wget -nc https://github.com/stolostron/policy-generator-plugin/releases/download/v1.9.1/linux-amd64-PolicyGenerator -O /tmp/linux-amd64-PolicyGenerator; chmod +x /tmp/linux-amd64-PolicyGenerator ; PolicyGenerator=/tmp/linux-amd64-PolicyGenerator ${GE ...
```
Except if this is added and you delete the already generated files it doesn't work.  I didn't dig into why.  But if you keep without `shell` it doesn't include uid or gid, it's `--user :` and generated files are owned by `root`.

lnguyen1401 replied:
so for some reason, having `$(shell id -u):$(shell id -g)` prevented the line from running. it does seem to include uid&gid `--user 4202615:4202615`. could it be just my laptop ?

jewzaam commented on scripts/generate-policy-config.py:
```diff
@@ -0,0 +1,38 @@
+import sys
+
+#an array of directories you want to generate policies for. Please make sure ONLY the directories you want exist here. if there's entries here, delete it and add yours.
+directory = ['./deploy/rbac-permissions-operator-config',
+        './deploy/osd-cluster-admin'
+        ]
```
Will be a little easier to read in my opinion.

```suggestion
directory = [
    './deploy/rbac-permissions-operator-config',
    './deploy/osd-cluster-admin'
]
```

jewzaam commented on scripts/generate-policy-config.py:
```diff
@@ -0,0 +1,38 @@
+import oyaml as yaml
+import shutil
+import os
+import sys
```
Unused import.

```suggestion
```

jewzaam reviewed (changes requested):
Does not handle generation of Policy from SubjectPermissions.

--- 

File permissions for generated Policy.yaml needs sorted out, see suggestions.  I wouldn't block PR on this probably, the generation of Policy for SubjectPermission is the important thing to work on.

jewzaam reviewed (commented):
Created bug https://issues.redhat.com/browse/OSD-14145 to track ownership issue when acm-policies files are generated from scratch (for me they're owned by root:root).

We have discussed the above issue and decide to track separately so we can move forward with this PR given it's only an impact when the files are first created.

jewzaam commented:
/approve
/lgtm

jewzaam commented:
/hold cancel
